File: 1424-maximum-candies-you-can-get-from-boxes/1424-maximum-candies-you-can-get-from-boxes.java
class Solution {
    public int maxCandies(int[] status, int[] candies, int[][] keys, int[][] containedBoxes, int[] initialBoxes) {
        int candy = 0;
        Queue<Integer> openBoxes = new LinkedList<>();
        LinkedList<Integer> lockedBoxes = new LinkedList<>();

        for (int box : initialBoxes) {
            if (status[box] == 1) {
                openBoxes.offer(box);
            } else {
                lockedBoxes.add(box);
            }
        }

        while (!openBoxes.isEmpty()) {
            while (!openBoxes.isEmpty()) {
                int box = openBoxes.poll();
                candy += candies[box];
                for (int key : keys[box]) {
                    status[key] = 1;
                }
                for (int containedBox : containedBoxes[box]) {
                    if (status[containedBox] == 1) {
                        openBoxes.offer(containedBox);
                    } else {
                        lockedBoxes.add(containedBox);
                    }
                }
            }
            Iterator<Integer> it = lockedBoxes.iterator();
            while (it.hasNext()) {
                int box = it.next();
                if (status[box] == 1) {
                    openBoxes.offer(box);
                    it.remove();
                }
            }
        }
        return candy;
    }
}

File: 1424-maximum-candies-you-can-get-from-boxes/1424-maximum-candies-you-can-get-from-boxes.cpp
class Solution {
public:
    int maxCandies(vector<int>& status, vector<int>& candies, vector<vector<int>>& keys, vector<vector<int>>& containedBoxes, vector<int>& initialBoxes) {
        int n = status.size(), candy = 0;
        vector<int> boxes;
        boxes.resize(n);
        for (int i = 0; i < n; i++) {
            boxes[i] = 0;
        }
        for (int i = 0; i < initialBoxes.size(); i++) {
            boxes[initialBoxes[i]] = 1;
        }
        
        for (int i = 0; i < initialBoxes.size(); i++) {
            int box = initialBoxes[i];
            if (status[box] && boxes[box]) {
                status[box] = 0;
                boxes[box] = 0;
                candy += openBox(box, status, boxes, candies, keys, containedBoxes);
                
            }
        }
        return candy;
    }
    int openBox (int box, vector<int>& status, vector<int>& boxes, vector<int>& candies, vector<vector<int>>& keys, vector<vector<int>>& containedBoxes) {
        int candy = candies[box];
        for (int j = 0; j < keys[box].size(); j++) {
            int box_ = keys[box][j];
            if (boxes[box_] == 1) {
                status[box_] = 0;
                boxes[box_] = 0;
                candy += openBox(box_, status, boxes, candies, keys, containedBoxes);
            } else {
                status[box_] = 1;
            }
        }
        for (int j = 0; j < containedBoxes[box].size(); j++) {
            int box_ = containedBoxes[box][j];
            if (status[box_] == 1) {
                status[box_] = 0;
                boxes[box_] = 0;
                candy += openBox(box_, status, boxes, candies, keys, containedBoxes);
            } else {
                boxes[box_] = 1;
            }
        }
        return candy;
    }
};